blueice/likelihood.py

```diff
@@ -101,29 +102,85 @@ def __init__(self, pdf_base_config, likelihood_config=None, **kwargs):
 
         # If there are shape parameters:
         self.anchor_models = OrderedDict()  # dictionary mapping model zs -> actual model
+        self.anchor_sources = OrderedDict()  # dictionary mapping source_name -> zs -> source
         # Interpolators created by morphers. These map zs to...
         self.mus_interpolator = None                          # rates for each source
         self.ps_interpolator = None                          # (source, event) p-values (unbinned), or pmf grid (binned)
         # number of events per bin observed in Monte Carlo / calibration data that gave rise to the model.
         self.n_model_events_interpolator = lambda x: None
         self.n_model_events = None
 
+    @property
+    def source_shape_parameters(self):
+        """Dict of sources with shape parameters. source name -> dict of shape parameters."""
+        source_shape_parameters = OrderedDict()
+        for sn, source, apply_eff, eff_name in zip(
+                self.source_name_list,
+                self.base_model.sources,
+                self.source_apply_efficiency,
+                self.source_efficiency_names
+                ):
+            ignore_parameters = set(source.config['dont_hash_settings'])
+            # The efficiecny parameter doesn't need to be hashed but it needs to be passed to the morpher
+            if apply_eff:
+                ignore_parameters.discard(eff_name)
+            shape_parameters = OrderedDict({k: v for k, v in self.shape_parameters.items() if k not in ignore_parameters})
+            if shape_parameters:
+                source_shape_parameters[sn] = shape_parameters
+        return source_shape_parameters
+
+    def _get_shape_indices(self, source_name):
+        """Return the indices of the shape parameters used by the source."""
+        shape_keys = self.source_shape_parameters[source_name].keys()
+        return [i for i, k in enumerate(self.shape_parameters.keys()) if k in shape_keys]
```

Comment:
for arrays should we return it sorted? Or return a set?

Reply:
Maybe I misunderstand your suggestion but I think the return should by this definition be a sorted list, or not? Also entries should be unique.